### ISSUE TITLE
SDK is raising an error when using variables inside fragments. 

### ISSUE THREAD

**Issue by thomasdelgado** [apollographql/apollo-ios#1191]
Hello guys, first of all thanks for the awesome library! You're doing an amazing job! :D

So, I'm having this problem here, I'm not sure If I can do something to work around it. 
I'm having trouble using variables inside fragments. The query works fine in playground, and still creates all the classes related, but it's raising a compile error. 

given the following query: 
```
query fetchProductSummary($productId: Int!, $salesChannel: String!) {
  getProduct(productId: $productId) {
    ...productSummary
  }
}


fragment productSummary on Product {
  id
  name
  thumbnailUrl
  prices(salesChannel: $salesChannel) {
    oldPrice
    price
  }
}
```

I'm getting the following error: 

> Apollo CLI Information: apollo/2.27.2 darwin-x64 node-v10.20.1
> + /Users/thomasdelgado/code//Pods/Apollo/scripts/apollo/bin/run codegen:generate --target=swift '--includes=./**/*.graphql' --localSchemaFile=schema.json API.swift
> Loading Apollo Project [started]
> Loading Apollo Project [completed]
> Generating query files [started]
> Generating query files with 'swift' target [title changed]
> file:///Users/thomasdelgado/code/TokStock-iOS/Tok%26Stock/App/GraphQL/Product.graphql:56: error: Unknown argument "salesChannel" on field "prices" of type "Product".
> file:///Users/thomasdelgado/code/TokStock-iOS/Tok%26Stock/App/GraphQL/ProductDetail.graphql:55: error: Unknown argument "salesChannel" on field "prices" of type "Product".
> Validation of GraphQL query document failed
>     at Object.validateQueryDocument (/Users/thomasdelgado/code/TokStock-iOS/Pods/Apollo/scripts/apollo/node_modules/apollo-language-server/lib/errors/validation.js:35:38)
>     at Object.generate [as default] (/Users/thomasdelgado/code/TokStock-iOS/Pods/Apollo/scripts/apollo/lib/generate.js:23:18)
>     at write (/Users/thomasdelgado/code/TokStock-iOS/Pods/Apollo/scripts/apollo/lib/commands/client/codegen.js:78:54)
>     at Task.task (/Users/thomasdelgado/code/TokStock-iOS/Pods/Apollo/scripts/apollo/lib/commands/client/codegen.js:98:46)
> Generating query files with 'swift' target - wrote 1 files [title changed]
> Generating query files with 'swift' target - wrote 1 files [completed]
> Command PhaseScriptExecution emitted errors but did not return a nonzero exit code to indicate failure

Full log: 
https://www.dropbox.com/s/nue96zypnr1xnz8/Screenshot%202020-05-05%2016.36.50.png?dl=0

This error is not blocking me from compiling in Xcode. But our CI server is not being able to create builds.

**Comment by designatednerd** [comment]
One big caveat I'll start with: Not everything that is perfectly legal in GraphQL and might work in a playground will work in a codegen environment. We try really hard to cover as much as possible, but there are certain types of flexibility - in this case, it appears part of that is treating `$salesChannel` as a global variable - that unfortunately it's extremely difficult to represent in a type safe fashion.

Basically, what it comes down to is that fragment doesn't really have a way of knowing, on its own, that the `$salesChannel` variable is going to be provided, and there's no way to create a fragment that takes a parameter. I'm honestly a little surprised this is generating code at all once it's hit that error. 

What does the generated code for the Fragment look like?

**Comment by thomasdelgado** [comment]
Thanks for the fast reply! 

I understand, we didn't know about that. I'm sending the code fragment. Another thing to point out is that this fragment was being used in multiples places without the variable, when we've added `salesChannel`, everything stayed the same without any problems (except for this compile error). 

```
public struct ProductSummary: GraphQLFragment {
  /// The raw GraphQL definition of this fragment.
  public static let fragmentDefinition: String =
    """
    fragment productSummary on Product {
      __typename
      id
      name
      thumbnailUrl
      prices(salesChannel: $salesChannel) {
        __typename
        oldPrice
        price
      }
    }
    """

  public static let possibleTypes: [String] = ["Product"]

  public static let selections: [GraphQLSelection] = [
    GraphQLField("__typename", type: .nonNull(.scalar(String.self))),
    GraphQLField("id", type: .scalar(Int.self)),
    GraphQLField("name", type: .scalar(String.self)),
    GraphQLField("thumbnailUrl", type: .scalar(String.self)),
    GraphQLField("prices", arguments: ["salesChannel": GraphQLVariable("salesChannel")], type: .object(Price.selections)),
  ]

  public private(set) var resultMap: ResultMap

  public init(unsafeResultMap: ResultMap) {
    self.resultMap = unsafeResultMap
  }

  public init(id: Int? = nil, name: String? = nil, thumbnailUrl: String? = nil, prices: Price? = nil) {
    self.init(unsafeResultMap: ["__typename": "Product", "id": id, "name": name, "thumbnailUrl": thumbnailUrl, "prices": prices.flatMap { (value: Price) -> ResultMap in value.resultMap }])
  }

  public var __typename: String {
    get {
      return resultMap["__typename"]! as! String
    }
    set {
      resultMap.updateValue(newValue, forKey: "__typename")
    }
  }

  public var id: Int? {
    get {
      return resultMap["id"] as? Int
    }
    set {
      resultMap.updateValue(newValue, forKey: "id")
    }
  }

  public var name: String? {
    get {
      return resultMap["name"] as? String
    }
    set {
      resultMap.updateValue(newValue, forKey: "name")
    }
  }

  public var thumbnailUrl: String? {
    get {
      return resultMap["thumbnailUrl"] as? String
    }
    set {
      resultMap.updateValue(newValue, forKey: "thumbnailUrl")
    }
  }

  public var prices: Price? {
    get {
      return (resultMap["prices"] as? ResultMap).flatMap { Price(unsafeResultMap: $0) }
    }
    set {
      resultMap.updateValue(newValue?.resultMap, forKey: "prices")
    }
  }

  public struct Price: GraphQLSelectionSet {
    public static let possibleTypes: [String] = ["ProductPrices"]

    public static let selections: [GraphQLSelection] = [
      GraphQLField("__typename", type: .nonNull(.scalar(String.self))),
      GraphQLField("oldPrice", type: .scalar(Double.self)),
      GraphQLField("price", type: .scalar(Double.self)),
    ]

    public private(set) var resultMap: ResultMap

    public init(unsafeResultMap: ResultMap) {
      self.resultMap = unsafeResultMap
    }

    public init(oldPrice: Double? = nil, price: Double? = nil) {
      self.init(unsafeResultMap: ["__typename": "ProductPrices", "oldPrice": oldPrice, "price": price])
    }

    public var __typename: String {
      get {
        return resultMap["__typename"]! as! String
      }
      set {
        resultMap.updateValue(newValue, forKey: "__typename")
      }
    }

    public var oldPrice: Double? {
      get {
        return resultMap["oldPrice"] as? Double
      }
      set {
        resultMap.updateValue(newValue, forKey: "oldPrice")
      }
    }

    public var price: Double? {
      get {
        return resultMap["price"] as? Double
      }
      set {
        resultMap.updateValue(newValue, forKey: "price")
      }
    }
  }
}```

**Comment by designatednerd** [comment]
Are all the other operations using this fragment now also providing `$salesChannel` variables?

**Comment by thomasdelgado** [comment]
Yes, we have 6 different queries using this fragment and all are providing the `$salesChannel` variable :)

**Comment by designatednerd** [comment]
@thomasdelgado Hey sorry this got lost in the ether - can you send me a small sample project with just your schema, the fragment this is using, and a query that uses that fragment? ellen at apollographql dot com. 

One other thing to double check before you do that is that your schema is totally up to date - it's possible that maybe something got out of sync in terms of parameters.

**Comment by thomasdelgado** [comment]
Hey @designatednerd! No worries! 

Actually, since we were in hurry, we've ended asking for the backend team to change this approach, so we could use the fragments without having to use variables. 

Because of that, I don't have that outdated schema :(

**Comment by designatednerd** [comment]
D'oh! Sorry I couldn't be of more timely help. Let me know if you want to continue trying to figure this out or if we should close this issue out for now.

**Comment by thomasdelgado** [comment]
It's all good! For me we close it. 
Have a great day!

**Comment by designatednerd** [comment]
Cool - anyone else having problems around this, please open a new issue, and I'll try to help you there. Thank you!
